### PR TITLE
3 think about ways how to fit functions of parameters

### LINKED ISSUE #3: Think about ways how to fit functions of parameters .
p(r),q(r), xi(r) etc,

### PR DESCRIPTION
**Pull request by Minijanach**
Simplified model was added, looks alright but the residual and the score look kind of weird to me. @prashjet please have a look at the gp_simplified_p.ipynb , i'm open for suggestions.

### CODE REVIEW

**Comment by prashjet** [comment]
I think this simplified version has helped! I played around with your notebook in the following steps:

- `gp_simplified_p_001` is the same as your notebook with the addition of a new residual plot at the end, where the errors are scaled by the mean prediction. Normalising the residuals this way, things look quite good!
- in `gp_simplified_p_002` I changed `AutoDelta` guide to `AutoNormal`.  It now looked like the predicted p value is 0.4 rather than 0.6... which is worrying... but oddly the image still looked quite reasonable. This suggested to me that there might be some mistake in the way we are interpreting the parameters of the guide...
- in `gp_simplified_p_003` I changed the process for sampling the posterior and posterior predictive. The predicted p value is now back to 0.6! That's good! But why!? I think the explanation is that...
   - When using a guide other than `AutoDelta`, internally, the guide parameters (e.g. `p_auto_loc`) are transformed to a different co-ordinate system (an _unconstrained_ space vs. the _constrained_ space that we have specified with our prior)
   - We therefore can't interpret the guide parameters directly, but instead we must extract posterior samples from the guide using `Predictive`. This undoes the transformation, taking us from unconstrained back to constrained space
   - Why does `AutoDelta` behave differently...? I'm not 100% sure, but it kind of makes sense that constraints can be handled differently for this case
   - I was then curious about the 'X' pattern in your residual plot. I thought this could be an artefact of the only varying `p` while leaving all other parameter fixed. To check this I ran one last test...
-  in `gp_simplified_p_004` I let `rho0` be free parameter. The X pattern is still there, but less strong. I guess that freeing other parameters (especially the viewing angles) would blur this out even further.

I think/hope this new process for sampling the posterior might solve the issues we were seeing when using GPs.

**Comment by Minijanach** [comment]
@prashjet After our last meeting we removed the sigmoid transform from the model and tried to do the same using the clip function. This changed the results in a more significant way than i expected (also it took a while to realize that it was because of the clipping). I prepared a comparison notebook for the clip and sigmoid models (compare_clip_sigmoid) and also tried it out for different input images (clip_for_different_inputs). Please take a look.
I also played around with different mean functions (linear,polynomial, two linear slopes and some others) these did not seem to have a big impact on the results (I initially thought they did until i realized it was because of the clip function).

**Comment by prashjet** [comment]
Hi @Minijanach - replacing the sigmoid transform with clip looks promising! At least in some cases we are now constraining p(r) correctly in the inner regions.  For the cases where our recovered p(r) still goes wrong in the inner regions, I still think there's some hope that a linear mean function could be helpful, with some changes to the current implementation.

I'd build this in two steps:

### 1.  a model with a purely linear p(r) i.e. no GP at all

I would parameterise this as follows:
```
p0 = numpyro.sample('p0', dist.Normal(0.5., 1.)) # i.e. = p(0)
p5 = numpyro.sample('p5', dist.Normal(0.5., 1.)) # i.e. = p(5)
p_not_transformed = p0 + (p5-p0)/5 * r
p = numpyro.deterministic("p", jnp.clip(p_not_transformed, min = 0.00001, max = 1.))
```

I think it's worth showing that this simplified model can produce something sensible before moving to the next step...

### 2.  a model with  p(r) = linear mean function + deviations from linear modelled with a GP

Try restricting the amplitude of the GP  (i.e. sigma_p) by decreasing the prior scale from 1 to 0.1 i.e.
```
sigma_p = numpyro.sample("sigma_p", dist.TruncatedNormal(0., 0.1 , low = 0.01, high=3.))
```
This will hopefully restrict the GP to modelling only the _deviations_ from linear. Previously, with a scale of 1, the GP had the flexibility to model the entire function, which might explain why you didn't see a big impact using the linear mean.

Could you re-run `clip_for_different_inputs.ipynb` with these two models? Keep the same input p(r)'s and SVI settings. If we see an improvement in the inner regions, great! If not, at least the results look decent enough now that we can move on to freeing the other functions q(r), xi(r), etc...

**Comment by Minijanach** [comment]
@prashjet So i added some new notebooks. clip_for_different_inputs_linear_nogp and clip_for_different_inputs_linear_gp we talked already about in the meeting.
I added two were i investiagted how the clip function affects the zero mean model with 300 and 500 steps (clip_for_different_inputs_zero_mean and clip_for_different_inputs_zero_mean_1000_steps).
Also one where we use the new parametrization (clip_for_different_inputs_linear_gp_different_parametrization), here we get again a loss that spikes to infinity for one of the images but the ones that run don't seem to have a significant improvement, i will take a look if i can get it to run completely.
The last one is using a zero mean and adding the mean function created with the new parametrization on top (clip_for_different_inputs_linear_gp_different_parametrization_added_to_mean) i see some improvement here for some cases especially the linear slopes but it gets worse for the const_high input. Please take a look if i implemented it as you had in mind.

**Comment by prashjet** [comment]
Thanks @Minijanach.  I think the re-parameterisation of of the linear mean function has been successful, since the recoveries now oscillate around the true  solution, rather than drop off to 0 at the end. Also a huge improvement increasing from 300 to 500 steps. I'd say stick to this combination going forward.

It's still confused why we get no uncertainties. I can believe that the uncertainties should be tiny within the range of the data, but the fact that there is no spread in recovered p(r) even beyond r>7 is slightly concerning. 

Perhaps we get some insight into uncertainties when we free other parameters. Before freeing q(r), xi(r), I'd say first try to freeing the viewing angles: give them priors in the model with some spread (say +/- 30 degrees) around their true values. If there's still no visible uncertainty in the results after that, then we might as well use an AutoDelta guide instead of Auto normal for the next steps.

**Comment by Minijanach** [comment]
@prashjet Changes as mentioned in the e-mail, please take a look!
